File: backend_py/store/mcp_tool_manifests.py
"""Persist MCP server tool lists (name + description) per user after successful connect."""

import json
from pathlib import Path

try:
    from config import DATA_DIR
except ImportError:
    DATA_DIR = Path(__file__).resolve().parent.parent / "data"

MANIFESTS_FILE = Path(DATA_DIR) / "mcp-tool-manifests.json"


def _read_all() -> dict:
    if not MANIFESTS_FILE.exists():
        return {}
    try:
        with open(MANIFESTS_FILE, encoding="utf-8") as f:
            data = json.load(f)
        return data if isinstance(data, dict) else {}
    except Exception:
        return {}


def _write_all(stored: dict) -> None:
    MANIFESTS_FILE.parent.mkdir(parents=True, exist_ok=True)
    with open(MANIFESTS_FILE, "w", encoding="utf-8") as f:
        json.dump(stored, f, indent=2)


def save_mcp_tool_manifest(user_id: str, server_id: str, tools: list[dict]) -> None:
    stored = _read_all()
    if user_id not in stored:
        stored[user_id] = {}
    stored[user_id][server_id] = {
        "tools": [{"name": t.get("name", ""), "description": t.get("description", "")} for t in tools],
    }
    _write_all(stored)


def get_mcp_tool_manifests(user_id: str) -> dict[str, list[dict]]:
    stored = _read_all()
    user = stored.get(user_id)
    if not user or not isinstance(user, dict):
        return {}
    return {
        sid: (m.get("tools") or [])
        for sid, m in user.items()
        if isinstance(m, dict) and "tools" in m
    }

File: backend/src/store/mcp-tool-manifests.ts
/**
 * Persist MCP server tool lists (name + description) per user after successful connect.
 * Stored in data/mcp-tool-manifests.json.
 */

import { readFile, writeFile, mkdir } from 'fs/promises';
import { join } from 'path';
import { config } from '../config.js';

const MANIFESTS_FILE = join(config.dataDir, 'mcp-tool-manifests.json');

type ToolEntry = { name: string; description?: string };
type ServerManifest = { tools: ToolEntry[] };
type Stored = Record<string, Record<string, ServerManifest>>;

async function ensureDir(): Promise<void> {
  await mkdir(config.dataDir, { recursive: true });
}

async function readAll(): Promise<Stored> {
  try {
    const data = await readFile(MANIFESTS_FILE, 'utf-8');
    const parsed = JSON.parse(data);
    return typeof parsed === 'object' && parsed !== null ? parsed : {};
  } catch {
    return {};
  }
}

async function writeAll(stored: Stored): Promise<void> {
  await ensureDir();
  await writeFile(MANIFESTS_FILE, JSON.stringify(stored, null, 2), 'utf-8');
}

export async function saveMcpToolManifest(
  userId: string,
  serverId: string,
  tools: Array<{ name: string; description?: string }>
): Promise<void> {
  const stored = await readAll();
  if (!stored[userId]) stored[userId] = {};
  stored[userId][serverId] = {
    tools: tools.map((t) => ({ name: t.name, description: t.description ?? '' })),
  };
  await writeAll(stored);
}

export async function getMcpToolManifests(userId: string): Promise<Record<string, ToolEntry[]>> {
  const stored = await readAll();
  const user = stored[userId];
  if (!user || typeof user !== 'object') return {};
  const out: Record<string, ToolEntry[]> = {};
  for (const [serverId, manifest] of Object.entries(user)) {
    if (manifest?.tools) out[serverId] = manifest.tools;
  }
  return out;
}
